datafusion/src/optimizer/filter_push_down.rs

```diff
@@ -409,10 +420,15 @@ fn optimize(plan: &LogicalPlan, mut state: State) -> Result<LogicalPlan> {
             //
             // Join clauses with `Using` constraints also take advantage of this logic to make sure
             // predicates reference the shared join columns are pushed to both sides.
```

Comment:
I think you are on the right track here that the filter_push_down is not correct for outer joins 

However, I don't think the ability to push filters is related to if there is an `IS NULL` predicate. The code should NEVER try to push predicates for outer join

So schematically, given this input:

```
Filter: A.a > 5
  Inner Join(A.a = B.b)
    Scan A
    Scan B
```

The code is effectively both pushing down `A.a > 5` as well as adding a *new* predicate, `B.b > 5` during the scan:
```
Filter: A.a > 5
  Inner Join(A.a = B.b)
    Scan A (A.a > 5) 
    Scan B (B.b > 5)
```

However, this transformation is not correct for outer joins where only the predicates on the preserved side (A in this case) can be pushed down. Thus for

```
Filter: A.a > 5
  LEFT OUTER Join(A.a = B.b)
    Scan A
    Scan B
```

We should still push `A.a > 5` but not introduce `B.b > 5`:
```
Filter: A.a > 5
  LEFT OUTER Join(A.a = B.b)
    Scan A (A.a > 5)
    Scan B
```